### PR TITLE
Add DNS or GW for use by dnsmasq

### PR DESCRIPTION
**Pull request by shibaPuppy**
Depending on your environment, you should be able to use either GW or DNS.

### CODE REVIEW

**Review by elfosardo — COMMENTED** [review]
thanks, this is ok, I think we should also document the variables in https://github.com/metal3-io/metal3-docs/blob/main/docs/user-guide/src/ironic/ironic_variables.md

**Comment by shibaPuppy** [comment]
@elfosardo 
https://github.com/metal3-io/metal3-docs/pull/289

File: ironic-config/dnsmasq.conf.j2 [hunk 1]
```diff
@@ -12,8 +12,19 @@ log-dhcp
 dhcp-range={{ env.DHCP_RANGE }}
 
 # Disable default router(s) and DNS over provisioning network
+# It can be used when setting DNS or GW variables.
+{%- if env["GW_IP"] is undefined %}
```
Comment (dtantsur) [inline ironic-config/dnsmasq.conf.j2:16]: Please don't shorten words: `GATEWAY_IP`

File: ironic-config/dnsmasq.conf.j2 [hunk 2]
```diff
@@ -12,8 +12,19 @@ log-dhcp
 dhcp-range={{ env.DHCP_RANGE }}
 
 # Disable default router(s) and DNS over provisioning network
```
Comment (dtantsur) [inline ironic-config/dnsmasq.conf.j2:14]: This comment should be moved

**Comment by dtantsur** [comment]
In addition to the user guide, let's also update the README in this repo?

File: ironic-config/dnsmasq.conf.j2 (hunk 2 above)
Comment (dtantsur) [reply ironic-config/dnsmasq.conf.j2:14]: ... or even removed completely since you have another comment just below

**Comment by dtantsur** [comment]
I'd be also interested in being able to use `DNS_IP=IRONIC_IP`, but I can look into that in a follow-up.

**Comment by dtantsur** [comment]
Something like

```diff
diff --git a/scripts/rundnsmasq b/scripts/rundnsmasq
index 7a4fa45..09ce32b 100755
--- a/scripts/rundnsmasq
+++ b/scripts/rundnsmasq
@@ -8,6 +8,10 @@ export DNS_PORT=${DNS_PORT:-0}
 
 wait_for_interface_or_ip
 
+if [[ "${DNS_IP:-}" == "provisioning" ]]; then
+    export DNS_IP=$IRONIC_IP
+fi
+
 mkdir -p /shared/tftpboot
 mkdir -p /shared/tftpboot/arm64-efi
 mkdir -p /shared/html/images
```

File: scripts/rundnsmasq [hunk 3]
```diff
@@ -7,6 +7,10 @@ DNSMASQ_EXCEPT_INTERFACE=${DNSMASQ_EXCEPT_INTERFACE:-"lo"}
 export DNS_PORT=${DNS_PORT:-0}
 
 wait_for_interface_or_ip
+if [[ "${DNS_IP:-}" == "provisioning" ]]; then
+    export DNS_IP=$IRONIC_IP
+fi
+
```
Comment (shibaPuppy) [inline scripts/rundnsmasq:13]: @dtantsur 
added it, but I don't know if it can act as an 'ironic' dns.

File: ironic-config/dnsmasq.conf.j2 (hunk 2 above)
Comment (shibaPuppy) [reply ironic-config/dnsmasq.conf.j2]: remove that comment.

**Comment by shibaPuppy** [comment]
@dtantsur 
is it my broken?

**Comment by dtantsur** [comment]
/test-ubuntu-integration-main

No, it's an upstream Ironic breakage. A fix has been merged, we need to wait until it gets into repositories.

**Comment by dtantsur** [comment]
/test-ubuntu-integration-main

The last attempt before I merge a CI workaround.

**Comment by shibaPuppy** [comment]
@dtantsur 
thx,
may i ask review below?
https://github.com/metal3-io/baremetal-operator/pull/1167#issue-1377547395

**Pull request merged**